### ISSUE TITLE
Cannot set property 'lazy_count' of undefined

### ISSUE THREAD

**Issue by rscholant** [hgourvest/node-firebird#249]
## Cannot set property 'lazy_count' of undefined

After performing some operations on the database, this error is occurring and the application is finishing running.
The strangest thing is that this error happens sporadically, there is no way to predict it.

### Possible solution

After some tests, it was observed that in the line shown below, a value is being assigned to the lazy_count property in which it is undefined.

### index.js - Line 3036
```javascript
if (err) {
    xdr.buffer = xdr.buffer.slice(pos);
    xdr.pos = 0;
    self._xdr = xdr;

    if (self.accept.protocolMinimumType === ptype_lazy_send && self._queue.length > 0) {
        self._queue[0].lazy_count = 2; //THIS LINE IS THE PROBLEM
    }
    return;
}
```

With this change the error no longer occurs, but the return of the query is strange ...
```javascript
if (err) {
    xdr.buffer = xdr.buffer.slice(pos);
    xdr.pos = 0;
    self._xdr = xdr;

    if (self.accept.protocolMinimumType === ptype_lazy_send 
            && self._queue.length > 0 
            && self._queue[0] //NEW CONDITION
            && self._queue[0].lazy_count  //NEW CONDITION ) {
        self._queue[0].lazy_count = 2;
    }
    return;
}
```

**Comment by iget-master** [comment]
This issue is happening here too with some queries

**Comment by stafocher** [comment]
Did you manage to solve it?

**Comment by jesusvilla** [comment]
Hi @rscholant
Can you share your code and the steps to follow to reproduce the error please?

**Comment by rhengles** [comment]
@jesusvilla 

> Hi @rscholant Can you share your code and the steps to follow to reproduce the error please?

Not OP, but I got the exact same error only by doing something extremely simple like this:

```js
const query = (db, sql) => new Promise((resolve, reject) =>
	db.query(sql, (err, result) =>
		err ? reject(err) : resolve(result)
	)
)

// later
sqlCount = `SELECT COUNT(*) AS "count" FROM SAME_TABLE`
sqlPage = `SELECT * FROM SAME_TABLE ROWS 1 TO 20`
const [count, page] = await Promise.all([
	query(db, sqlCount),
	query(db, sqlPage)
])
```

**Comment by rodrigo-rafael** [comment]
i have the same lazy_count problem

**Comment by hcrufio88** [comment]
The problem is still going on. I have the issue on a large query that fetch 1000-1500 results, each with a blob field.
I have the blobastext option active in the db options with firebird 2.5.
Any suggestions?

**Comment by ivanseidel** [comment]
same problem here!

**Comment by vvvait** [comment]
у меня зависало на попытке загрузить из базы исходники процедур, удавалось загрузить примерно 150 штук из 300+
я уменьшил размер буфера до 80 и стали загружаться все, но правда в 10+ раз медленнее
```
Connection.prototype.getSegment = function(blob, callback) {
    var msg = this._msg;
    msg.pos = 0;
    msg.addInt(op_get_segment);
    msg.addInt(blob.handle);
    //msg.addInt(1024); // buffer length    
    msg.addInt(80); // <---- buffer length    
    msg.addInt(0); // ???
    this._queueEvent(callback);
};
```
Думаю лучше не использовать blobAsText = true из-за этого:

`infoValue.value += data.toString(DEFAULT_ENCODING);`

если будут многобайтные символы, то они попадают в разные пакеты и вы получите битый текст.

и этого:

```
const arrPromise = (ret.arrBlob || []).map(value => value(transaction));
Promise.all(arrPromise).then((arrBlob) => {
```

блобы лучше грузить не разом все, а по 10 за раз

**Comment by slachtar** [comment]
Interesting, I'm not using blobasText option because of this and I'm doing something else to fix the issue.
I'm following this issue and I think it should have high priority because all of us use blobs and can't get them properly.

I'm transalating what you wrote using google translate (think to translate next time because we don't, at least me, understand Russian). 
....

I was stuck trying to download the source codes of procedures from the database, I was able to download about 150 out of 300+
I reduced the buffer size to 80 and everything started loading, but it’s actually 10+ times slower

```
Connection.prototype.getSegment = function(blob, callback) {
     var msg = this._msg;
     msg.pos = 0;
     msg.addInt(op_get_segment);
     msg.addInt(blob.handle);
     //msg.addInt(1024); // buffer length
     msg.addInt(80); // <---- buffer length
     msg.addInt(0); // ???
     this._queueEvent(callback);
};
```
I think it's better not to use blobAsText = true because of this:

`infoValue.value += data.toString(DEFAULT_ENCODING);`

if there are multibyte characters, then they end up in different packets and you will get broken text.

and this:

```
const arrPromise = (ret.arrBlob || []).map(value => value(transaction));
Promise.all(arrPromise).then((arrBlob) => {
```
It’s better to load blobs not all at once, but 10 at a time

**Comment by vvvait** [comment]
я нашел принципиальную ошибку в загрузке блобов, когда их много и они больше 1024 байта firebird начинает их раскладывать по разным фрэймам, а это не учитывается при разборе пакета.
Исправленная версия в моём форке, также там добавлена настройка fetchSize и segmentSize указывается в custom
(если увеличить segmentSize  c 1K до 8K то время выполнения запроса с текстовым блобом [select RDB$PROCEDURE_SOURCE from RDB$PROCEDURES] уменьшается с 680мс до 250мс, но это зависит ещё от размера блобов)
есть возможность получить имена полей даже для пустого датасета.
пример:
```
query(db: Database, tr: Transaction|undefined, sql: string, params: any[] = [], custom = {}) {
    return new Promise<IQueryResult>((resolve, reject) => {
      setTimeout(() => reject(new Error("Таймаут выполнения запроса")), 12 * DEFAULT_TIMEOUT);
      let columns: any = null;
      const onColumns = (cols: any) => columns = cols.filter((f: string) => !f.startsWith("__"));
      (db as any).once("columns", onColumns);
      const cb = (err: any, rows: any[]|any|undefined) => {
        (db as any).off("columns", onColumns);
        if (err) {
          error("query", err, sql, params);
          reject(err);
        } else {
          if (!Array.isArray(rows))
            resolve({ columns, returning: rows });
          else {
            resolve({ columns, rows });
          }
        }
      };
      debug(sql, params);
      if (tr)
        tr.execute(sql, params, cb, { asObject: true, asStream: false, ...custom });
      else 
        db.query(sql, params, cb);
    });
  }
```

**Comment by ekoncept** [comment]
Szanowni Państwo;Do 16 kwietnia 2024 r. jestem poza dostępem do skrzynki email. Wszelkie bieżące sprawy proszę kierować na ***@***.*** lub do Pana Marka Latarskiego. Ja odpowiem dopiero po przyjeździe. Pozdrawiam;Grzegorz Krzyżanowski

**Comment by vvvait** [comment]
Проблема что указана в начале темы происходит когда вместе с командой серверу не указывается колбэк, а конкретно 
здесь:
`statement.connection.closeBlob(blob);`
если ответ на этот запрос не приходит полностью в одном фрейме а точнее в одном буфере data
`... _socket.on('data', function(data) { ...`
происходит обработка ошибки, и ожидается следующий буфер

**Comment by felipedllinfo** [comment]
Hi, 
I am using the example of the lib. Howerver in sometimes give me this msg. In another msgs just don't returning me anything, just timeout. When I use with breakpoint The error doenst appears. When i use in my query first 7. the error doesn't happens too. but after 7, 8, 9 .. sometimes happesn another doesn't.

return new Promise(async (resolve, reject) => {
        try {
            Firebird.attach(conexao, (err, db) => {
                if (err)
                    throw err;

                db.transaction(Firebird.ISOLATION_READ_COMMITTED, (err, transaction) => {
                    if (err) {
                        throw err;
                    }

                    transaction.query(sql, params, (err, result) => {
                        if (err) {
                            transaction.rollback();
                            reject(err)
                            return;
                        }

                        const arrBlob = [];
                        for (const item of result) {
                            const fields = Object.keys(item);
                            for (const key of fields) {
                                if (typeof item[key] === 'function') {
                                    item[key] = new Promise((resolve, reject) => {
                                        item[key](transaction, (error, name, event, row) => {
                                            if (error) {
                                                reject(error);
                                                return
                                            }

                                            let buffers = [];
                                            event.on('data', (chunk) => {
                                                buffers.push(chunk)
                                            });
                                            event.on('end', () => {
                                                let data = Buffer.concat(buffers)
                                                data = data.toString('base64')

                                                resolve({ value: data, column: name, row });
                                            });
                                        });
                                    });
                                    arrBlob.push(item[key]);
                                }
                            }
                        }

                        Promise.all(arrBlob).then((blobs) => {
                            for (const blob of blobs) {
                                result[blob.row][blob.column] = blob.value;
                            }

                            transaction.commit((err) => {
                                if (err) {
                                    transaction.rollback();
                                    reject(err)
                                    return;
                                }

                                db.detach();
                                resolve(result)
                            });
                        }).catch((err) => {
                            transaction.rollback();
                            reject(err)
                            return
                        });
                    });
                });
            });
        } catch (error) {
            reject(error)
        }
    })

**Comment by mariuz** [comment]
This crash class is fixed in the 2.x line. `lazy_count of undefined` was the visible symptom of the response queue desynchronizing around lazy-protocol deferred operations: every deferred packet still gets a server response, and the queue used to pair those responses with the wrong callbacks (sporadically, depending on packet timing — exactly the unpredictability you saw). The 2.x releases fixed the deferred-response accounting (placeholders per deferred op), the trailing `op_execute2` response handling, and stale per-packet decode state — with regression tests for each. Closing — reopen with a stack if anything similar appears on 2.10+.
